mllib/src/main/scala/org/apache/spark/ml/nlp/lbfgs.scala

```diff
@@ -0,0 +1,281 @@
+        }
+        i = 0
+        while (i < size) {
+          w(i) = diag(i) * w(i)
+          i += 1
+        }
+        i = 0
+        while (i < bound) {
+          val yr: Double = ddot(size, w, iypt + cp * size + 1, w, 1)
+          var beta: Double = w(size + cp + 1) * yr
+          val inmc: Int = size + msize + cp
+          beta = w(inmc) - beta
+          iscn = ispt + cp * size
+          daxpy(size, beta, w, iscn, w, 0)
+          cp += 1
+          if (cp == msize) cp = 0
+          i += 1
+        }
+        i = 0
+        while (i < size) {
+          w(ispt + point * size + i) = w(i)
+          i += 1
+        }
+      }
+      // mcsrch line search
+      if (info != -1) {
+        infoc = 1
+        if (size <= 0 || stp <= 0.0) {
+          return
+        }
+        dginit = ddot(size, g, 0, w, ispt)
+        if (dginit >= 0.0) return
+        brackt = false
+        stage1 = true
+        nfev = 0
+        finit = f
+        dgtest = ftol * dginit
+        width = 1e20 - 1e-20
+        width1 = width / p5
+        j = 0
+        while (j < size) {
+          diag(j) = x(j)
+          j += 1
+        }
+        stx = 0.0
+        fx = finit
+        dgx = dginit
+        sty = 0.0
+        fy = finit
+        dgy = dginit
+      }
+      while (persist) {
+        if (info != -1) {
+          stmin = stx
+          stmax = stp + xtrapf * (stp - stx)
+          stp = math.max(stp, 1e-20)
+          stp = math.min(stp, 1e20)
+          j = 0
+          while (j < size) {
+            x(j) = diag(j) + stp * w(j + ispt)
+            j += 1
+          }
+          info = -1
+          persist = false
+        } else {
+          info = 0
+          nfev += 1
+          val dg: Double = ddot(size, g, 0, w, ispt)
+          val ftest1: Double = finit + stp * dgtest
+
+          if (stp == 1e20 && f <= ftest1 && dg <= dgtest) {
+            info = 5
+          }
+          if (stp == 1e-20 && (f > ftest1 || dg >= dgtest)) {
+            info = 4
+          }
+          if (nfev >= maxfev) {
+            info = 3
+          }
+          if (f <= ftest1 && math.abs(dg) <= 0.9 * (-dginit)) {
+            info = 1
+          }
+          if (info != 0) {
+            persist = false
+          }
+        }
+      }
+      // mcsrch
+      if (info == -1) {
+        iflag = 1
+        return
+      }
+      if (info != 1) {
+        iflag = -1
+        return
+      }
+      npt = point * size
+      i = 0
+      while (i < size) {
+        w(ispt + npt + i) = stp * w(ispt + npt + i)
+        w(iypt + npt + i) = g(i) - w(i)
+        i += 1
+      }
+      point += 1
+      if (point == msize) point = 0
+      val gnorm: Double = math.sqrt(ddot(size, v, 0, v, 0))
+      val xnorm: Double = math.max(1.0, math.sqrt(ddot(size, x, 0, x, 0)))
+      if (gnorm / xnorm <= eps) {
+        iflag = 0
+      }
+    }
+  }
+
+  def ddot(size: Int, v1: ArrayBuffer[Double], v1start: Int,
+           v2: ArrayBuffer[Double], v2start: Int): Double = {
```

Comment:
You can use `org.apache.spark.mllib.linalg.Vector` type and Spark provides a class that provides BLAS functionality. (`org.apache.spark.mllib.linalg.BLAS`) with this Vector type. It might be bette to avoid re-implement it for `ArrayBuffer`.